riscv/src/register/macros.rs

```diff
@@ -600,3 +600,510 @@ macro_rules! clear_pmp {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: [$bit_start:literal : $bit_end:literal]$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start <= $bit_end);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> usize {
+                $crate::bits::bf_extract(self.bits, $bit_start, $bit_end - $bit_start + 1)
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     $(#[$field_ty_doc:meta])+
+     $field_ty:ident {
+         range: [$field_start:literal : $field_end:literal],
+         default: $default_variant:ident,
+         $($variant:ident = $value:expr$(,)?)+
+     }$(,)?
+    ) => {
+        $crate::csr_field_enum!(
+            $(#[$field_ty_doc])+
+            $field_ty {
+                range: [$field_start : $field_end],
+                default: $default_variant,
+                $($variant = $value,)+
+            },
+        );
+
+        $crate::read_only_csr_field!(
+            $ty,
+            $(#[$field_doc])*
+            $field,
+            $(#[$try_field_doc])*
+            $try_field,
+            $field_ty,
+            range: [$field_start : $field_end],
+        );
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     $field_ty:ident,
+     range: [$field_start:literal : $field_end:literal]$(,)?
+    ) => {
+        const _: () = assert!($field_end < usize::BITS);
+        const _: () = assert!($field_start <= $field_end);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> $field_ty {
+                self.$try_field().unwrap()
+            }
+
+            $(#[$try_field_doc])+
+            #[inline]
+            pub fn $try_field(&self) -> $crate::result::Result<$field_ty> {
+                let value = $crate::bits::bf_extract(
+                    self.bits,
+                    $field_start,
+                    $field_end - $field_start + 1,
+                );
+
+                $field_ty::from_usize(value)
+            }
+        }
+    };
+}
+
+/// Defines field accesor functions for a write-only CSR type.
+#[macro_export]
+macro_rules! write_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit:literal$(,)?) => {
+        const _: () = assert!($bit < usize::BITS);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&mut self, $field: bool) {
+                self.bits = $crate::bits::bf_insert(self.bits, $bit, 1, $field as usize);
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     range: $bit_start:literal..=$bit_end:literal$(,)?) => {
```

Comment:
Same here:

```suggestion
     range: $bit_start:literal: $bit_end:literal$(,)?) => {
```

Reply:
Same as above, this branch is to define a range of single-bit fields.